Add support for posting interactive messages

Rukas opened the pull request:
Hi,

I've created a few commands related to posting interactive messages to Slack (https://github.com/RamblingCookieMonster/PSSlack/issues/37).  This also modifies `New-SlackMessageAttachment` to include the necessary parameters for adding actions. I also had to update `Send-SlackMessage` to set the depth on `ConvertTo-Json` to account for the new actions/options. Please check it out and let me know what you think.

RamblingCookieMonster commented:
Awesome : D  cc @devblackops 

Things are a bit busy, might take me a little time, but will read up and play around with it as soon as I have time!

RamblingCookieMonster commented:
Cool, got to play with some azure functions : )

So!  This is all way new to me.  A couple suggestions, take these with a grain of salt : )

* Is there a case where someone would indicate an action, without a callbackid?  If not, might be worth adding some indication (or potentially not running) if an action is specified but callbackid isn't bound (again - I might just be missing common or uncommon cases where this would happen).
* Example code has a few spots that didn't work for how I tested this (same caveat as above):
  * CallbackID not specified
  * New-SlackMessage channel not specified (this might be due to me using a token vs. a uri)
  * Some spot with a missing escape ``` at the end - or I copied/pasted wrong : P  will mark it here if I find it

That said - it works : D  Thanks!  I'll leave this open for another day or two in case @devblackops has any input

Cheers!

RamblingCookieMonster commented on PSSlack/Public/New-SlackMessageAttachment.ps1:
```diff
@@ -170,12 +181,39 @@ function New-SlackMessageAttachment
+        # Construct and send the message!
+        New-SlackMessageAttachment -Color $_PSSlackColorMap.orange `
+                                   -Title 'Failed to process account' `
+                                   -Actions $action
```
ah, here it is - `

RamblingCookieMonster commented on PSSlack/Public/New-SlackActionConfirmation.ps1:
```diff
@@ -0,0 +1,90 @@
+        # Construct and send the message!
+        New-SlackMessageAttachment -Color $_PSSlackColorMap.orange `
+                                   -Title 'Failed to process account' `
+                                   -Actions $action
```
another ` here

devblackops commented:
I haven't had a chance to play with this yet but I'll take a look 👍

Rukas commented:
You are correct that an `Action` will require a `CallbackId` and vice versa.  I've added a new parameter set in `New-SlackMessageAttachment` to account for this.  I also updated the examples with the missing escape characters, thanks for catching those!

From my understanding, actions will require that a web hook be used and will not work with a token since they are dependent on a Slack app to execute the actions.  The only way I could think of validating this in `Send-SlackMessage` is to peek at the attachment(s) and see if there are any actions attached.  If there are then make sure a web hook URI is passed and not a token

RamblingCookieMonster commented:
Awesome, thanks!

Seems to work with tokens as well!  Here's what I did:

* Create an app (https://api.slack.com/apps)
* Enable Interactivity, Pick a Request URL (Interactive Components)
* Add a bot user to the app
* Install the app in your space
* Copy the token (Permissions / OAuth and Permissions - I think I used xoxb-*)
* Play

Cheers!

RamblingCookieMonster commented:
I've skimmed the docs and code again - will merge tonight unless there are any objections - worst case scenario we can change things later.

Thanks again, this is awesome!

devblackops commented on PSSlack/Public/New-SlackMessageAttachment.ps1:
```diff
@@ -223,6 +261,15 @@ function New-SlackMessageAttachment
         [System.Collections.Hashtable[]]$Fields,
+
+        [Parameter(Mandatory=$true,
+                    ParameterSetName='Action')]
```
@Rukas Is this parameter set needed? When specifying an action, you can't also specify `Color` or `Severity` because they're in different parameter sets.

Rukas replied:
That's a good point. I used that as an easy way to make sure that a callback ID is specified when an action is specified. I guess I could add a `BEGIN` block and have it validate that both were specified and throw if one is but the other isn't.  Any thoughts on a better way?

RamblingCookieMonster replied:
Good catch Brandon!

Yeah, just logic to validate that they're together, and throw if not might make the most sense?  Probably more readable / less complex than fun with multiple parametersets (e.g. color-action, color, severity-action, severity, or whatever combination we'd need to get them both in place), and should still be usable enough.  What do you all think?

Rukas replied:
I've added the changed specified above. Let me know what you think!

RamblingCookieMonster commented:
Alrighty, looks good to me, worst case scenario we can tweak later : ) Thanks again, this is awesome!
